Connectors/Rhino/Speckle.Connectors.GrasshopperShared/Components/Objects/CreateSpeckleBlockDefinition.cs

```diff
@@ -0,0 +1,164 @@
+
+    // TODO: what about description, base point parameter, color/Material overrides for the definition itself
+  }
+
+  protected override void RegisterOutputParams(GH_OutputParamManager pManager)
+  {
+    pManager.AddParameter(
+      new SpeckleBlockDefinitionWrapperParam(),
+      "Block Definition",
+      "BD",
+      "Speckle Block Definition",
+      GH_ParamAccess.item
+    );
+
+    pManager.AddParameter(
+      new SpeckleObjectParam(),
+      "Objects",
+      "O",
+      "Objects contained in the Block Definition",
+      GH_ParamAccess.list
+    );
+
+    pManager.AddTextParameter("Name", "N", "Name of the Block Definition", GH_ParamAccess.item);
+  }
+
+  protected override void SolveInstance(IGH_DataAccess da)
+  {
+    // keep track of mutation
+    bool mutated = false;
+
+    // are we modifying an existing block definition or creating a new one?
+    SpeckleBlockDefinitionWrapperGoo? inputBlockDef = null;
+    da.GetData(0, ref inputBlockDef);
+
+    SpeckleBlockDefinitionWrapper result;
+    if (inputBlockDef != null) // if != null → user has piped in a definition and we're modifying existing
+    {
+      result = inputBlockDef.Value.DeepCopy();
+    }
+    else // if == null → we're creating a brand spanking new one
+    {
+      result = new SpeckleBlockDefinitionWrapper()
+      {
+        Base = new InstanceDefinitionProxy
+        {
+          name = "Unnamed Block",
+          objects = new List<string>(),
+          maxDepth = 1
+        },
+        Objects = new List<SpeckleObjectWrapper>(),
+        ApplicationId = Guid.NewGuid().ToString()
+      };
+      mutated = true;
+    }
+
+    // get whatever objects the user wants inside the block definition
+    List<SpeckleObjectWrapperGoo> inputObjects = new();
+    da.GetDataList(1, inputObjects);
+
+    if (inputObjects.Count > 0)
+    {
+      var processedObjects = new List<SpeckleObjectWrapper>();
+      var objectIds = new List<string>();
+
+      foreach (var objGoo in inputObjects)
+      {
+        if (objGoo?.Value != null)
+        {
+          var obj = objGoo.Value.DeepCopy();
+
+          // ensure the object has an application ID
+          obj.ApplicationId ??= Guid.NewGuid().ToString();
+          processedObjects.Add(obj);
+          objectIds.Add(obj.ApplicationId);
+        }
+      }
+
+      result.Objects = processedObjects; // update objects
+      result.InstanceDefinitionProxy.objects = objectIds;
+      mutated = true;
+    }
+
+    // name for the block
+    string? inputName = null;
+    da.GetData(2, ref inputName);
+
+    if (inputName != null)
+    {
+      result.Name = inputName;
+      result.InstanceDefinitionProxy.name = inputName;
+      mutated = true;
+    }
+
+    if (mutated)
+    {
+      result.ApplicationId = Guid.NewGuid().ToString();
+      result.InstanceDefinitionProxy.applicationId = result.ApplicationId;
+    }
+
+    // we need a valid name
+    if (string.IsNullOrEmpty(result.Name))
+    {
```

Comment:
this may be a bit dangerous on send - we should probably generate a meaningful application id